multicore_runtime/middleware.py

```diff
@@ -0,0 +1,185 @@
+
+    return app(wsgi_env, start_response)
+
+  return reset_environment_wrapper
+
+
+def request_environment_for_wsgi_env(wsgi_env):
+  """Get a dict of key-value pairs from wsgi_env that work as env variables.
+
+  Does not mutate input.
+
+  Args:
+    wsgi_env: WSGI request env data.
+
+  Returns:
+    A dictionary suitable for use in `os.environ.update(output)`.
+  """
+
+  # Return all key, value pairs from wsgi_env where value is a string.
+  return {key: value for key, value in wsgi_env.iteritems()
+          if isinstance(value, basestring)}
+
+
+def reserved_env_keys_for_wsgi_env(wsgi_env):
+  """Get a dict for reserved keys based on wsgi_env headers and defaults.
+
+  Does not mutate input.
+
+  Args:
+    wsgi_env: WSGI request env data.
+
+  Returns:
+    A dictionary suitable for use in `os.environ.update(output)`.
+  """
+
+  output = {}
+
+  # Use the default value for a reserved key if the corresponding header is not
+  # set, or if the header exists but its value is blank.
+  for key, default in RESERVED_ENV_KEYS.iteritems():
+    value = wsgi_env.get('HTTP_X_APPENGINE_' + key)
+    output[key] = value or default  # Must be set to a valid value or default.
+
+  return output
+
+
+def get_env_to_hide_service_bridge(wsgi_env):
+  """Generate a dictionary of environment variables to hide the service bridge.
+
+  Does not mutate input.
+
+  Args:
+    wsgi_env: WSGI request env data.
+
+  Returns:
+    A dictionary suitable for use in `os.environ.update(output)`.
+  """
+  output = {}
+
+  # Because the request is coming over the service bridge, the service bridge
+  # host and port that are automatically populated in SERVER_NAME and
+  # SERVER_PORT, respectively.  However, this is an implementation detail that
+  # should not be shown to user code. Instead, we'll rely on the HTTP Host
+  # header to retrieve the hostname used in the original request.  This mimics
+  # the behavior of non-VM App Engine.
+  if 'HTTP_HOST' in wsgi_env:
+    output['SERVER_NAME'] = wsgi_env['HTTP_HOST']
+
+  # Similarly we'll use the HTTPS flag to determine the port used in the
+  # original request.
+  https = wsgi_env.get('HTTP_X_APPENGINE_HTTPS', 'off')
+  if https == 'off':
+    output['SERVER_PORT'] = '80'
+  elif https == 'on':
+    output['SERVER_PORT'] = '443'
+  else:
+    logging.warning('Unrecognized value for HTTPS (%s), won\'t modify '
+                    'SERVER_PORT', https)
+
+  return output
+
+
+def health_check_middleware(app):
+  """Intercept requests to /_ah/health and respond healthy (200 OK).
+
+  Args:
+    app: The WSGI app to wrap.
+
+  Returns:
+    The wrapped app, also a WSGI app.
+  """
+
+  def health_check_intercept_wrapper(wsgi_env, start_response):  # pylint: disable=missing-docstring
+    request = Request(wsgi_env)
+    if request.path == '/_ah/health':  # Only intercept exact matches.
+      return Response('healthy', status=200)(wsgi_env, start_response)
```

Comment:
I think it's slightly more clear if you assign the Response() to a variable and then call it later.

See my comment in `dispatcher.py` about perhaps reusing the `app` variable for this purpose, in which case you can reuse a single `return app(...)` statement for both cases.